R/workhorse.R

```diff
@@ -27,57 +28,79 @@ build_request <- function(endpoint = NULL,
 ## right now not setting endpoint, if you want to include parameters, must be in a named list
 ## may introduce the :parameter notation in endpoint later :)
 
-set_query <- function(x){
-  if (length(x$params) == 0L) return(x)
+set_query <- function(x) {
+  if (length(x$params) == 0L)
+    return(x)
   if (x$method != "GET") {
     if (grepl("\\?", x$endpoint)) {
       x$query <- sub(".*\\?", "", x$endpoint)
       return(x)
     }
     return(x)
   }
-  if (!all(has_names(x$params))){
+  if (!all(has_names(x$params))) {
     spf("All parameters must be named.")
   }
   x$query <- x$params
   x$params <- NULL
   x
 }
 
-set_body <- function(x){
-  if (length(x$params) == 0L) return(x)
+set_body <- function(x) {
+  if (length(x$params) == 0L)
+    return(x)
 
   x$body <- x$params
   x
 }
 
 ## not setting headers yet
 
-set_url <- function(x){
-
-  if (grepl("^http", x$endpoint)){
+set_url <- function(x) {
+  if (grepl("^http", x$endpoint)) {
     x$url <- x$endpoint
   } else{
     x$url <- file.path(x$api_url, sub(".", "", x$endpoint))
   }
   x
 }
 
-make_request <- function(x, ...){
-  method <-  list("GET" = httr::GET,
-                  "POST" = httr::POST,
-                  "PATCH" = httr::PATCH,
-                  "PUT" = httr::PUT,
-                  "DELETE" = httr::DELETE)[[x$method]]
-  method(url = x$url,
-         x$token,
-         query = x$query,
-         body = x$body, ...)
+make_request <- function(x, ...) {
+  method <-  list(
+    "GET" = httr::GET,
+    "POST" = httr::POST,
+    "PATCH" = httr::PATCH,
+    "PUT" = httr::PUT,
+    "DELETE" = httr::DELETE
+  )[[x$method]]
+  method(
+    url = x$url,
+    x$token,
+    query = x$query,
+    body = x$body,
+    ...
+  )
 }
 
-process_request <- function(res, content = TRUE) {
+process_request <- function(res,
+                            content = TRUE,
+                            expected = "application/json; charset=UTF-8",
+                            as = "parsed",
+                            encoding = "UTF-8") {
   httr::stop_for_status(res)
-  if (content == TRUE){
-    httr::content(res)
+  if (content == TRUE) {
+    actual <- res$headers$`content-type`
+    if (actual != expected) {
+      spf(
+        paste0(
+          "Expected content-type:\n%s",
+          "\n",
+          "Actual content-type:\n%s"
+        ),
+        expected,
+        actual
+      )
+    }
+    httr::content(res, as = as, encoding = encoding)
```

Comment:
I think this is *safe*: you're checking that content declares itself as UTF-8 JSON then you let `httr::content()` auto-parse. I think it is slightly unconventional? For example, if you read the docs for `httr::content()`, `as = "parsed"` is discouraged in a package. But, as I said, I think you are doing the necessary checks.